com.unity.multiplayer.mlapi/Tests/Runtime/NetworkVariableHelper.cs

```diff
@@ -0,0 +1,154 @@
+using System;
+using System.Collections.Generic;
+using MLAPI.NetworkVariable;
+
+namespace MLAPI.RuntimeTests
+{
+    /// <summary>
+    /// Will automatically register for the NetworkVariable OnValueChanged
+    /// delegate handler.  It then will expose that single delegate invocation
+    /// to anything that registers for this NetworkVariableHelper's instance's OnValueChanged event.
+    /// This allows us to register any NetworkVariable type as well as there are basically two "types of types":
+    /// IEquatable<T>
+    /// ValueType
+    /// From both we can then at least determine if the value indeed changed
+    /// </summary>
+    /// <typeparam name="T"></typeparam>
+    internal class NetworkVariableHelper<T> : BaseNetworkVariableHelper
+    {
+        private NetworkVariable<T> m_NetworkVariable;
+        public delegate void OnMyValueChangedDelegateHandler(T previous, T next);
+        public event OnMyValueChangedDelegateHandler OnValueChanged;
+
+
+        /// <summary>
+        /// IEquatable<T> Equals Check
+        /// </summary>
+        private void CheckVariableChanged(IEquatable<T> previous, IEquatable<T> next)
+        {
+            if (!previous.Equals(next))
+            {
+                ValueChanged();
+            }
+        }
+
+        /// <summary>
+        /// ValueType Equals Check
+        /// </summary>
+        private void CheckVariableChanged(ValueType previous, ValueType next)
+        {
+            if (!previous.Equals(next))
+            {
+                ValueChanged();
+            }
+        }
+
+        /// <summary>
+        /// INetworkVariable's OnVariableChanged delegate callback
+        /// </summary>
+        /// <param name="previous"></param>
+        /// <param name="next"></param>
+        private void OnVariableChanged(T previous, T next)
+        {
+            var testValueType = previous as ValueType;
+            if (testValueType != null)
+            {
+                CheckVariableChanged(previous as ValueType, next as ValueType);
+            }
+            else
+            {
+                CheckVariableChanged(previous as IEquatable<T>, next as IEquatable<T>);
+            }
+        }
+
+        public NetworkVariableHelper(INetworkVariable networkVariable) : base(networkVariable)
+        {
+            m_NetworkVariable = networkVariable as NetworkVariable<T>;
+            m_NetworkVariable.OnValueChanged = OnVariableChanged;
+        }
+    }
+
+    /// <summary>
+    /// The BaseNetworkVariableHelper keeps track of:
+    /// The number of instances and associates the instance with the NetworkVariable
+    /// The number of times a specific NetworkVariable instance had its value changed (i.e. !Equal)
+    /// Note: This could be expanded for future tests focuses around NetworkVariables
+    /// </summary>
+    internal class BaseNetworkVariableHelper
+    {
+        private static Dictionary<BaseNetworkVariableHelper, INetworkVariable> s_Instances;
```

Comment:
What would it take to have less(no?) static content? static makes scaling complex, as the singleton removal showed. I can see the point of "we want a single test for all variables", but I wonder, what if we ended up with two such tests running in parallel in some future, would the static come back to bite us?

Reply:
I like that thought...so when you create a generic typed class it effectively creates a unique instance of the class for each unique type applied to the generic class.  Since I am storing the instances of both the BaseNetworkVariableHelper and a reference to the INetworkVariable instance, there should be no collision unless the instances for both client and server are exactly the same.  My assumption here is that the server-host and any client(s) being tested would not have the exact same instances for both class types. 
So two reasons:

1. Creating a local-non-static version of the dictionary would yield many instances of the dictionary for each unique NetworkVariableHelper<T> instance... which that would be problematic determining total counts etc.
2. The values being stored --should-- all be unique, and as such there shouldn't be the kind of collision you described.